cypress/e2e/shared/fluxQueryBuilder.test.ts

```diff
@@ -344,7 +345,99 @@ describe('Script Builder', () => {
         )
       })
 
-      it('should clear the editor text, and schema browser, with a new script', () => {
+      describe('conditions for divergence:', () => {
+        it('diverges when typing in composition block', () => {
+          cy.getByTestID('flux-sync--toggle').should('have.class', 'active')
+          cy.getByTestID('flux-editor', {timeout: 30000})
+          selectSchema()
+          confirmSchemaComposition()
+
+          cy.log('does not diverge when outside block')
+          cy.getByTestID('flux-editor').monacoType('// will not diverge')
+          cy.getByTestID('flux-sync--toggle').should(
+            'not.have.class',
+            'disabled'
+          )
+
+          cy.log('does diverge, within block')
+          cy.getByTestID('flux-editor').monacoType(
+            '{enter}{upArrow}{upArrow}{upArrow} // make diverge'
+          )
+          cy.log('toggle is now disabled')
+          cy.getByTestID('flux-sync--toggle').should('have.class', 'disabled')
+        })
+
+        describe('using hotkeys:', () => {
+          const runTest = (hotKeyCombo: string) => {
+            cy.getByTestID('flux-sync--toggle').should('have.class', 'active')
+            cy.getByTestID('flux-editor', {timeout: 30000})
+            selectSchema()
+            confirmSchemaComposition()
+
+            cy.log('does not diverge when outside block')
+            cy.getByTestID('flux-editor').monacoType(`foo ${hotKeyCombo}`)
+            cy.getByTestID('flux-sync--toggle').should(
+              'not.have.class',
+              'disabled'
+            )
+
+            cy.log('does diverge, within block')
+            cy.getByTestID('flux-editor').monacoType(
+              `{upArrow}{upArrow}{upArrow} ${hotKeyCombo}`
+            )
+            cy.log('toggle is now disabled')
+            cy.getByTestID('flux-sync--toggle').should('have.class', 'disabled')
+          }
+
+          it('diverges when commenting line', () => {
+            runTest('{cmd}/')
+          })
+
+          it('diverges when adding lines', () => {
+            runTest('{shift+alt+downArrow}')
+          })
+
+          it('diverges when removing lines', () => {
+            runTest('{cmd+x}')
+          })
+
+          it('diverges when moving lines', () => {
+            runTest('{alt+downArrow}')
+          })
+
+          /// XXX: wiedld (27 Sep 2022) -- we have no way to delineate btwn the LSP applyEdit,
```

Comment:
excellent question.

Here are the order of sub-events, for a schema composition event:
1. executeCommand message is sent to the LSP
2. LSP returns an applyEdit (which we cannot see/detect directly)
3. monaco-editor applies the change
4. we can detect this change using the "change object"

Here are the order of sub-events, for a user triggered hotkey:
1. hotkey pressed
2. editor action is triggered
3. action method is executed
4. monaco-editor applies the change
5. we can then detect the "change object"

The change object has the following information:
* text to replace
* position start/end for replacement

### When we decide to diverge

The LSP applyEdits replace the entire composition block. So we use a heuristic which says that if the "change object" (1) replaces this text, and (2) matches an expected start/end range --> then it's probably a change in response to an LSP requested applyEdit.

Anything which does not match this heuristic --> is probably from the user --> diverge.

This heuristic breaks down with the undo/redo shortcut keys. Because the change object is EXACTLY the same as a previous LSP-triggered change. (Because it's just undoing the last change.)

### Can we fix it? Should we fix it?

The only way we know definitively if the change was requested from the LSP, is if we start tracking our executeCommand requests to the LSP and the LSP applyEdit response. A.k.a. the middleware idea.

But we have decided to not go that route, and to instead: (1) replace any need for "diverge" in Q4, and (2) eventually replace all of the LSP-editor wiring (hopefully Q4). Hence why the unhandled edge case is documented here instead...so we explicitly know what we are not fixing.